# Video Game Sales Intelligence Suite
## Notebook 01: Data Understanding and Cleaning

---

### 1 - Importing Libraries and Setup Environment

In [2]:
import numpy as np
import pandas as pd
import matplotlib as plt

### 2 - Load Raw Data

In [3]:
# Load the raw dataset
AC = pd.read_csv("../data/raw/application_categories.csv")
AD = pd.read_csv("../data/raw/application_developers.csv")
AG = pd.read_csv("../data/raw/application_genres.csv")
AP = pd.read_csv("../data/raw/application_platforms.csv")
APu = pd.read_csv("../data/raw/application_publishers.csv")
application = pd.read_csv("../data/raw/applications.csv")

C:\Users\lenovo\AppData\Local\Temp\ipykernel_18280\1288600162.py:7: DtypeWarning: Columns (0: required_age) have mixed types. Specify dtype option on import or set low_memory=False.
  application = pd.read_csv("../data/raw/applications.csv")


#### the AI suggested multiple ways to load the datasets on of them is to ot put each dataset in a different variable
#### but I think it is better to put all the datasets in one variable and then we can use it to clean the data and 
#### then we can use it to analyze the data.


In [4]:
# Load additional datasets
categories = pd.read_csv("../data/raw/categories.csv")
developers = pd.read_csv("../data/raw/developers.csv")
genres = pd.read_csv("../data/raw/genres.csv")
platforms = pd.read_csv("../data/raw/platforms.csv")
publishers = pd.read_csv("../data/raw/publishers.csv")
reviews = pd.read_csv("../data/raw/reviews.csv")

# Combine all dataframes into a dictionary for easier management
dfs_dict = {
    'application_categories': AC,
    'application_developers': AD,
    'application_genres': AG,
    'application_platforms': AP,
    'application_publishers': APu,
    'applications': application,
    'categories': categories,
    'developers': developers,
    'genres': genres,
    'platforms': platforms,
    'publishers': publishers,
    'reviews': reviews
}

# Display basic info about each dataset
for name, df in dfs_dict.items():
    print(f"\n{name}:")
    print(f"  Shape: {df.shape}")
    print(f"  Columns: {list(df.columns)}")



application_categories:
  Shape: (1077156, 2)
  Columns: ['appid', 'category_id']

application_developers:
  Shape: (250817, 2)
  Columns: ['appid', 'developer_id']

application_genres:
  Shape: (587515, 2)
  Columns: ['appid', 'genre_id']

application_platforms:
  Shape: (334671, 2)
  Columns: ['appid', 'platform_id']

application_publishers:
  Shape: (223048, 2)
  Columns: ['appid', 'publisher_id']

applications:
  Shape: (239664, 30)
  Columns: ['appid', 'name', 'type', 'is_free', 'release_date', 'required_age', 'short_description', 'supported_languages', 'header_image', 'background', 'metacritic_score', 'recommendations_total', 'mat_supports_windows', 'mat_supports_mac', 'mat_supports_linux', 'mat_initial_price', 'mat_final_price', 'mat_discount_percent', 'mat_currency', 'mat_achievement_count', 'mat_pc_os_min', 'mat_pc_processor_min', 'mat_pc_memory_min', 'mat_pc_graphics_min', 'mat_pc_os_rec', 'mat_pc_processor_rec', 'mat_pc_memory_rec', 'mat_pc_graphics_rec', 'created_at', 'upd

##### if len(AC) > 0:: This checks if the table (e.g., Categories) actually has data in it. If it’s empty, it skips the merge to avoid errors.
##### suffixes=('', '_cat'): If both tables have columns with the same name (like a "name" column), Pandas needs to distinguish them. This tells Pandas to leave the master column name alone and add _cat to the end of the new column.

In [5]:
# Start with the main applications table as the base
combined_data = application.copy()

# Merge with application-relationship tables (these have appid)
# Merge application_categories
if len(AC) > 0:
    combined_data = combined_data.merge(AC, on='appid', how='left', suffixes=('', '_cat'))

# Merge application_developers
if len(AD) > 0:
    combined_data = combined_data.merge(AD, on='appid', how='left', suffixes=('', '_dev'))

# Merge application_genres
if len(AG) > 0:
    combined_data = combined_data.merge(AG, on='appid', how='left', suffixes=('', '_gen'))

# Merge application_platforms
if len(AP) > 0:
    combined_data = combined_data.merge(AP, on='appid', how='left', suffixes=('', '_plat'))

# Merge application_publishers
if len(APu) > 0:
    combined_data = combined_data.merge(APu, on='appid', how='left', suffixes=('', '_pub'))

print("Combined data shape:", combined_data.shape)
print("\nFirst 5 rows:")
print(combined_data.head())


Combined data shape: (4953412, 35)

First 5 rows:
   appid            name  type  is_free release_date required_age  \
0     10  Counter-Strike  game    False   2000-11-01            0   
1     10  Counter-Strike  game    False   2000-11-01            0   
2     10  Counter-Strike  game    False   2000-11-01            0   
3     10  Counter-Strike  game    False   2000-11-01            0   
4     10  Counter-Strike  game    False   2000-11-01            0   

                                   short_description  \
0  Play the world's number 1 online action game. ...   
1  Play the world's number 1 online action game. ...   
2  Play the world's number 1 online action game. ...   
3  Play the world's number 1 online action game. ...   
4  Play the world's number 1 online action game. ...   

                                 supported_languages  \
0  English<strong>*</strong>, French<strong>*</st...   
1  English<strong>*</strong>, French<strong>*</st...   
2  English<strong>*</strong>, 

#### Dataset Overview - Columns and Data Types


In [6]:
# Display data types and info about the combined dataset
print("Data Types:")
print(combined_data.dtypes)
print("\n" + "="*80)
print("\nDataset Info:")
print(combined_data.info())


Data Types:
appid                      int64
name                         str
type                         str
is_free                     bool
release_date                 str
required_age              object
short_description            str
supported_languages          str
header_image                 str
background                   str
metacritic_score         float64
recommendations_total    float64
mat_supports_windows        bool
mat_supports_mac            bool
mat_supports_linux          bool
mat_initial_price        float64
mat_final_price          float64
mat_discount_percent     float64
mat_currency                 str
mat_achievement_count    float64
mat_pc_os_min                str
mat_pc_processor_min         str
mat_pc_memory_min            str
mat_pc_graphics_min          str
mat_pc_os_rec                str
mat_pc_processor_rec         str
mat_pc_memory_rec            str
mat_pc_graphics_rec          str
created_at                   str
updated_at                   st

In [7]:
combined_data.head()

,appid,name,type,is_free,release_date,required_age,short_description,supported_languages,header_image,background,...,mat_pc_processor_rec,mat_pc_memory_rec,mat_pc_graphics_rec,created_at,updated_at,category_id,developer_id,genre_id,platform_id,publisher_id
0,10,Counter-Strike,game,False,2000-11-01,0,Play the world's number 1 online action game. ...,"English<strong>*</strong>, French<strong>*</st...",https://shared.akamai.steamstatic.com/store_it...,https://store.akamai.steamstatic.com/images/st...,...,NaN,NaN,NaN,2025-09-07 16:27:12.218587+00:00,2025-09-29 02:01:37.107239+00:00,72.0,59202.0,122.0,1,50182.0
1,10,Counter-Strike,game,False,2000-11-01,0,Play the world's number 1 online action game. ...,"English<strong>*</strong>, French<strong>*</st...",https://shared.akamai.steamstatic.com/store_it...,https://store.akamai.steamstatic.com/images/st...,...,NaN,NaN,NaN,2025-09-07 16:27:12.218587+00:00,2025-09-29 02:01:37.107239+00:00,72.0,59202.0,122.0,2,50182.0
2,10,Counter-Strike,game,False,2000-11-01,0,Play the world's number 1 online action game. ...,"English<strong>*</strong>, French<strong>*</st...",https://shared.akamai.steamstatic.com/store_it...,https://store.akamai.steamstatic.com/images/st...,...,NaN,NaN,NaN,2025-09-07 16:27:12.218587+00:00,2025-09-29 02:01:37.107239+00:00,72.0,59202.0,122.0,3,50182.0
3,10,Counter-Strike,game,False,2000-11-01,0,Play the world's number 1 online action game. ...,"English<strong>*</strong>, French<strong>*</st...",https://shared.akamai.steamstatic.com/store_it...,https://store.akamai.steamstatic.com/images/st...,...,NaN,NaN,NaN,2025-09-07 16:27:12.218587+00:00,2025-09-29 02:01:37.107239+00:00,90.0,59202.0,122.0,1,50182.0
4,10,Counter-Strike,game,False,2000-11-01,0,Play the world's number 1 online action game. ...,"English<strong>*</strong>, French<strong>*</st...",https://shared.akamai.steamstatic.com/store_it...,https://store.akamai.steamstatic.com/images/st...,...,NaN,NaN,NaN,2025-09-07 16:27:12.218587+00:00,2025-09-29 02:01:37.107239+00:00,90.0,59202.0,122.0,2,50182.0


#### Add Category Names
combined_data = combined_data.merge(lookup_categories, left_on='category_id', right_on='id', how='left', suffixes=('', '_cat_name'))

#### Add Developer Names
combined_data = combined_data.merge(lookup_developers, left_on='developer_id', right_on='id', how='left', suffixes=('', '_dev_name'))

#### Add Genre Names
combined_data = combined_data.merge(lookup_genres, left_on='genre_id', right_on='id', how='left', suffixes=('', '_gen_name'))

#### Add Platform Names
combined_data = combined_data.merge(lookup_platforms, left_on='platform_id', right_on='id', how='left', suffixes=('', '_plat_name'))

#### Add Publisher Names
combined_data = combined_data.merge(lookup_publishers, left_on='publisher_id', right_on='id', how='left', suffixes=('', '_pub_name'))

In [8]:
combined_data.head()

,appid,name,type,is_free,release_date,required_age,short_description,supported_languages,header_image,background,...,mat_pc_processor_rec,mat_pc_memory_rec,mat_pc_graphics_rec,created_at,updated_at,category_id,developer_id,genre_id,platform_id,publisher_id
0,10,Counter-Strike,game,False,2000-11-01,0,Play the world's number 1 online action game. ...,"English<strong>*</strong>, French<strong>*</st...",https://shared.akamai.steamstatic.com/store_it...,https://store.akamai.steamstatic.com/images/st...,...,NaN,NaN,NaN,2025-09-07 16:27:12.218587+00:00,2025-09-29 02:01:37.107239+00:00,72.0,59202.0,122.0,1,50182.0
1,10,Counter-Strike,game,False,2000-11-01,0,Play the world's number 1 online action game. ...,"English<strong>*</strong>, French<strong>*</st...",https://shared.akamai.steamstatic.com/store_it...,https://store.akamai.steamstatic.com/images/st...,...,NaN,NaN,NaN,2025-09-07 16:27:12.218587+00:00,2025-09-29 02:01:37.107239+00:00,72.0,59202.0,122.0,2,50182.0
2,10,Counter-Strike,game,False,2000-11-01,0,Play the world's number 1 online action game. ...,"English<strong>*</strong>, French<strong>*</st...",https://shared.akamai.steamstatic.com/store_it...,https://store.akamai.steamstatic.com/images/st...,...,NaN,NaN,NaN,2025-09-07 16:27:12.218587+00:00,2025-09-29 02:01:37.107239+00:00,72.0,59202.0,122.0,3,50182.0
3,10,Counter-Strike,game,False,2000-11-01,0,Play the world's number 1 online action game. ...,"English<strong>*</strong>, French<strong>*</st...",https://shared.akamai.steamstatic.com/store_it...,https://store.akamai.steamstatic.com/images/st...,...,NaN,NaN,NaN,2025-09-07 16:27:12.218587+00:00,2025-09-29 02:01:37.107239+00:00,90.0,59202.0,122.0,1,50182.0
4,10,Counter-Strike,game,False,2000-11-01,0,Play the world's number 1 online action game. ...,"English<strong>*</strong>, French<strong>*</st...",https://shared.akamai.steamstatic.com/store_it...,https://store.akamai.steamstatic.com/images/st...,...,NaN,NaN,NaN,2025-09-07 16:27:12.218587+00:00,2025-09-29 02:01:37.107239+00:00,90.0,59202.0,122.0,2,50182.0


In [9]:
cols_to_drop = [col for col in combined_data.columns if col == 'id']
combined_data = combined_data.drop(columns=cols_to_drop)

# Rename the 'name' columns to be more descriptive
combined_data = combined_data.rename(columns={
    'name': 'category_name',
    'name_dev_name': 'developer_name',
    'name_gen_name': 'genre_name',
    'name_plat_name': 'platform_name',
    'name_pub_name': 'publisher_name'
})

In [10]:
print("Final Table Preview:")
combined_data.head()

Final Table Preview:


,appid,category_name,type,is_free,release_date,required_age,short_description,supported_languages,header_image,background,...,mat_pc_processor_rec,mat_pc_memory_rec,mat_pc_graphics_rec,created_at,updated_at,category_id,developer_id,genre_id,platform_id,publisher_id
0,10,Counter-Strike,game,False,2000-11-01,0,Play the world's number 1 online action game. ...,"English<strong>*</strong>, French<strong>*</st...",https://shared.akamai.steamstatic.com/store_it...,https://store.akamai.steamstatic.com/images/st...,...,NaN,NaN,NaN,2025-09-07 16:27:12.218587+00:00,2025-09-29 02:01:37.107239+00:00,72.0,59202.0,122.0,1,50182.0
1,10,Counter-Strike,game,False,2000-11-01,0,Play the world's number 1 online action game. ...,"English<strong>*</strong>, French<strong>*</st...",https://shared.akamai.steamstatic.com/store_it...,https://store.akamai.steamstatic.com/images/st...,...,NaN,NaN,NaN,2025-09-07 16:27:12.218587+00:00,2025-09-29 02:01:37.107239+00:00,72.0,59202.0,122.0,2,50182.0
2,10,Counter-Strike,game,False,2000-11-01,0,Play the world's number 1 online action game. ...,"English<strong>*</strong>, French<strong>*</st...",https://shared.akamai.steamstatic.com/store_it...,https://store.akamai.steamstatic.com/images/st...,...,NaN,NaN,NaN,2025-09-07 16:27:12.218587+00:00,2025-09-29 02:01:37.107239+00:00,72.0,59202.0,122.0,3,50182.0
3,10,Counter-Strike,game,False,2000-11-01,0,Play the world's number 1 online action game. ...,"English<strong>*</strong>, French<strong>*</st...",https://shared.akamai.steamstatic.com/store_it...,https://store.akamai.steamstatic.com/images/st...,...,NaN,NaN,NaN,2025-09-07 16:27:12.218587+00:00,2025-09-29 02:01:37.107239+00:00,90.0,59202.0,122.0,1,50182.0
4,10,Counter-Strike,game,False,2000-11-01,0,Play the world's number 1 online action game. ...,"English<strong>*</strong>, French<strong>*</st...",https://shared.akamai.steamstatic.com/store_it...,https://store.akamai.steamstatic.com/images/st...,...,NaN,NaN,NaN,2025-09-07 16:27:12.218587+00:00,2025-09-29 02:01:37.107239+00:00,90.0,59202.0,122.0,2,50182.0


In [11]:
combined_data.describe()

,appid,metacritic_score,recommendations_total,mat_initial_price,mat_final_price,mat_discount_percent,mat_achievement_count,category_id,developer_id,genre_id,platform_id,publisher_id
count,4.953412e+06,191109.000000,7.971990e+05,3.386303e+06,3.386303e+06,3.386303e+06,1.611089e+06,4.920026e+06,4.929659e+06,4.846045e+06,4.953412e+06,4.439384e+06
mean,1.879878e+06,75.870676,8.588043e+03,3.425434e+04,3.186298e+04,6.049244e+00,3.727116e+01,1.687976e+02,4.896964e+04,6.334183e+01,1.440079e+00,4.186164e+04
std,1.102721e+06,9.440940,6.619369e+04,1.028839e+06,9.563143e+05,1.882122e+01,1.622321e+02,9.399237e+01,2.937516e+04,4.927045e+01,7.097678e-01,2.470676e+04
min,1.000000e+01,6.000000,1.010000e+02,5.000000e+01,4.400000e+01,0.000000e+00,0.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00
25%,8.979000e+05,71.000000,2.050000e+02,2.990000e+02,2.990000e+02,0.000000e+00,1.100000e+01,1.210000e+02,2.249900e+04,3.100000e+01,1.000000e+00,2.048600e+04
50%,1.783760e+06,77.000000,5.010000e+02,6.990000e+02,5.990000e+02,0.000000e+00,2.100000e+01,1.420000e+02,4.942800e+04,3.200000e+01,1.000000e+00,4.149000e+04
75%,2.808630e+06,82.000000,1.923000e+03,1.479000e+03,1.249000e+03,0.000000e+00,3.800000e+01,2.010000e+02,7.446900e+04,1.130000e+02,2.000000e+00,6.392000e+04
max,3.996190e+06,97.000000,4.661011e+06,8.990000e+07,8.910000e+07,9.500000e+01,9.821000e+03,5.330000e+02,1.015880e+05,1.670000e+02,3.000000e+00,8.600400e+04


The appid column has approximately $4.95 \times 10^6$ entries, while metacritic_score only has $1.91 \times 10^5$.
### This indicates a high volume of null values in the scores column.

In columns like mat_initial_price, the maximum value is exponentially higher than the 75% mark.
### This suggests the presence of outliers or potential data entry errors that could skew statistical models

In metacritic_score, the median (77) is quite close to the 75th percentile (82), indicating that the data is concentrated toward higher ratings.
### This helps determine if the data is normally distributed or skewed

Columns such as developer_id, genre_id, and platform_id show mean and standard deviation values.
### Since these are categorical identifiers (IDs) and not actual numerical measurements, these statistics are mathematically irrelevant. This serves as a reminder to convert these columns to object or category types to avoid logical errors in modeling.

The std (Standard Deviation) row measures how spread out the numbers are
High variance in recommendations_total shows a massive gap in popularity across the dataset
### High standard deviation often suggests that feature scaling (like Normalization or Standardization) may be necessary before training a model